velox/type/DecimalUtil.h

```diff
@@ -147,6 +148,55 @@ class DecimalUtil {
     return static_cast<TOutput>(rescaledValue);
   }
 
+  /// Rescale a double value to decimal value.
+  ///
+  /// Use `folly::tryTo` to convert a double value to int128_t or int64_t. It
+  /// returns an error when overflow occurs so that we could determine whether
+  /// an overflow occurs through checking the result.
+  ///
+  /// Normally, return the rescaled value. Otherwise, if the `toValue` overflows
+  /// the TOutput's limits or the `toValue` exceeds the precision's limits, it
+  /// will throw an exception.
+  template <typename TOutput>
+  inline static TOutput
+  rescaleDouble(double inputValue, const int toPrecision, const int toScale) {
+    VELOX_USER_CHECK(
+        std::isfinite(inputValue),
+        "Cannot cast {} '{}' to DECIMAL({},{})",
+        SimpleTypeTrait<double>::name,
+        std::isnan(inputValue)
+            ? "NAN"
+            : (std::isinf(inputValue) ? "INFINITY"
+                                      : std::to_string(inputValue)),
+        toPrecision,
+        toScale);
+
+    auto toValue =
+        inputValue * static_cast<double>(DecimalUtil::kPowersOfTen[toScale]);
+
+    TOutput rescaledValue;
+    bool isOverflow = !std::isfinite(toValue);
```

Comment:
To check `isfinite` looks redundant, because it is handled in previous user check.

Reply:
Not redundant, previous user check is to check the `inputValue`, and here is to check `toValue`.

Reply:
Interesting. Do you have a test case that hits this code path?

Reply:
Yes, I add a test case for this code path when the `inputValue` is `std::numeric_limits<double>::max()`.
When the `inputValue` is `std::numeric_limits<double>::max()` and the `scale` is any number not less than 1, the `toValue` will be more than `std::numeric_limits<double>::max()` and be overflowed.